> Ho provato ad aggiungere un test perchè per testare serve un'importazione per Aliquota e il test già esistente non lo faceva. Però il test va in errore ma non capisco perchè, sapreste aiutarmi a come sistemarlo?

Grazie dell'aggiunta! Lo stack dell'errore preso da https://github.com/OCA/l10n-italy/actions/runs/11794311853/job/32851644723?pr=4434#step:8:893 è
<details><summary>Stack</summary>
<pre>

2024-11-12 09:30:21,508 528 ERROR odoo odoo.addons.l10n_it_fatturapa_in_rc.tests.test_fatturapa_in_rc: ERROR: TestInvoiceRC.test_01_xml_import
Traceback (most recent call last):
  File "/__w/l10n-italy/l10n-italy/l10n_it_fatturapa_in_rc/tests/test_fatturapa_in_rc.py", line 13, in setUp
    self._create_account()
  File "/__w/l10n-italy/l10n-italy/l10n_it_fatturapa_in_rc/tests/test_fatturapa_in_rc.py", line 22, in _create_account
    self.account_selfinvoice = account_model.create(
  File "<decorator-gen-745>", line 2, in create
  File "/opt/odoo/odoo/api.py", line 414, in _model_create_multi
    return create(self, [arg])
  File "/__w/l10n-italy/l10n-italy/l10n_it_account/models/account_account.py", line 34, in create
    return super().create(vals_list)
  File "<decorator-gen-136>", line 2, in create
  File "/opt/odoo/odoo/api.py", line 415, in _model_create_multi
    return create(self, arg)
  File "/opt/odoo/addons/mail/models/mail_thread.py", line 258, in create
    threads = super(MailThread, self).create(vals_list)
  File "<decorator-gen-68>", line 2, in create
  File "/opt/odoo/odoo/api.py", line 415, in _model_create_multi
    return create(self, arg)
  File "/opt/odoo/odoo/addons/base/models/ir_fields.py", line 670, in create
    recs = super().create(vals_list)
  File "<decorator-gen-15>", line 2, in create
  File "/opt/odoo/odoo/api.py", line 415, in _model_create_multi
    return create(self, arg)
  File "/opt/odoo/odoo/models.py", line 3984, in create
    records = self._create(data_list)
  File "/opt/odoo/odoo/models.py", line 4165, in _create
    cr.execute(
  File "/opt/odoo/odoo/sql_db.py", line 321, in execute
    res = self._obj.execute(query, params)
psycopg2.errors.UniqueViolation: duplicate key value violates unique constraint "account_account_code_company_uniq"
DETAIL:  Key (code, company_id)=(295000, 10) already exists.

</pre>
</details>

Da qui si vede che durante il `setUp` cerca di creare un conto avente codice **295000**, ma fallisce perché esiste già.

Questo succede perché la classe dei test è `SingleTransactionCase`: https://github.com/OCA/l10n-italy/blob/2da29dd2a42fd82958c3b5465f1e848eb278a147/l10n_it_fatturapa_in/tests/fatturapa_common.py#L18 quindi ogni test (`setUp` + test + `tearDown`) viene eseguito in un'unica transazione: finché c'è solo un test no problemo; però ora che hai aggiunto un altro test esegue di nuovo il `setUp`, cerca di creare di nuovo lo stesso conto e giustamente fallisce.

Per risolvere puoi trasformare l'attuale `setUp` in `setUpClass`, se possibile andrebbe fatto in un commit di refactoring precedente alla fix che stai proponendo.

--- a/l10n_it_fatturapa_in_rc/tests/test_fatturapa_in_rc.py
+++ b/l10n_it_fatturapa_in_rc/tests/test_fatturapa_in_rc.py
@@ -159,3 +159,29 @@ class TestInvoiceRC(FatturapaCommon):
         invoice_id = res.get("domain")[0][2][0]
         invoice = self.invoice_model.browse(invoice_id)
         self.assertTrue(len(invoice.invoice_line_ids) == 0)
+
+    def test_01_xml_import(self):
+        supplier = self.env["res.partner"].search([("vat", "=", "IT02780790107")])[0]
+        # import lines grouped by "Tax rate"
+        supplier.e_invoice_detail_level = "1"
+        res = self.run_wizard(
+            "test2", "IT01234567890_FPR04.xml", module_name="l10n_it_fatturapa_in_rc"
+        )
+        # Trigger pending computations
+        # as if they were done during the attachment import,
+        # in the UX this happens right before going back to the client.
+        # Otherwise account.move.line.rc field is computed on demand
+        # and its check on the context fails.
+        self.env(
+            context=dict(
+                active_model="fatturapa.attachment.in",
+            ),
+        ).flush_all()
+
+        invoice_id = res.get("domain")[0][2][0]
+        invoice = self.invoice_model.browse(invoice_id)
+        self.assertEqual(invoice.invoice_line_ids[0].name, "Riepilogo Aliquota 22.00")
+        self.assertEqual(invoice.invoice_line_ids[1].name, "Riepilogo Aliquota 0.00")
+        self.assertFalse(invoice.invoice_line_ids[0].rc)
+        self.assertTrue(invoice.invoice_line_ids[1].rc)
+        self.assertEqual(invoice.invoice_line_ids[0].tax_ids.name, "22% e-bill")

--- a/l10n_it_fatturapa_in_rc/tests/test_fatturapa_in_rc.py
+++ b/l10n_it_fatturapa_in_rc/tests/test_fatturapa_in_rc.py
@@ -5,21 +5,23 @@ from odoo.addons.l10n_it_fatturapa_in.tests.fatturapa_common import FatturapaCom
 
 
 class TestInvoiceRC(FatturapaCommon):
-    def setUp(self):
-        super().setUp()
-        self.invoice_model = self.env["account.move"]
-        self.invoice_line_model = self.env["account.move.line"]
-        self.partner_model = self.env["res.partner"]
-        self._create_account()
-        self._create_taxes()
-        self._create_journals()
-        self._create_rc_types()
-        self._create_rc_type_taxes()
-        self._create_fiscal_position()
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.invoice_model = cls.env["account.move"]
+        cls.invoice_line_model = cls.env["account.move.line"]
+        cls.partner_model = cls.env["res.partner"]
+        cls._create_account()
+        cls._create_taxes()
+        cls._create_journals()
+        cls._create_rc_types()
+        cls._create_rc_type_taxes()
+        cls._create_fiscal_position()
 
-    def _create_account(self):
-        account_model = self.env["account.account"]
-        self.account_selfinvoice = account_model.create(
+    @classmethod
+    def _create_account(cls):
+        account_model = cls.env["account.account"]
+        cls.account_selfinvoice = account_model.create(
             {
                 "code": "295000",
                 "name": "selfinvoice temporary",
@@ -27,46 +29,48 @@ class TestInvoiceRC(FatturapaCommon):
             }
         )
 
-    def _create_taxes(self):
-        tax_model = self.env["account.tax"]
-        self.tax_22ai = tax_model.create(
+    @classmethod
+    def _create_taxes(cls):
+        tax_model = cls.env["account.tax"]
+        cls.tax_22ai = tax_model.create(
             {
                 "name": "Tax 22% Purchase RC ITA",
                 "type_tax_use": "purchase",
                 "amount": 22,
-                "kind_id": self.env.ref("l10n_it_account_tax_kind.n6_1").id,
+                "kind_id": cls.env.ref("l10n_it_account_tax_kind.n6_1").id,
                 "sequence": 10,
             }
         )
-        self.tax_22vi = tax_model.create(
+        cls.tax_22vi = tax_model.create(
             {
                 "name": "Tax 22% Sales RC ITA",
                 "type_tax_use": "sale",
                 "amount": 22,
-                "kind_id": self.env.ref("l10n_it_account_tax_kind.n6_1").id,
+                "kind_id": cls.env.ref("l10n_it_account_tax_kind.n6_1").id,
                 "law_reference": "articoli 23 e 25 D.P.R. 633/1972",
                 "sequence": 10,
             }
         )
 
-    def _create_journals(self):
-        journal_model = self.env["account.journal"]
-        self.journal_selfinvoice = journal_model.create(
+    @classmethod
+    def _create_journals(cls):
+        journal_model = cls.env["account.journal"]
+        cls.journal_selfinvoice = journal_model.create(
             {
                 "name": "selfinvoice",
                 "type": "sale",
                 "code": "SLF",
             }
         )
-        self.journal_reconciliation = journal_model.create(
+        cls.journal_reconciliation = journal_model.create(
             {
                 "name": "RC reconciliation",
                 "type": "bank",
                 "code": "SLFRC",
-                "default_account_id": self.account_selfinvoice.id,
+                "default_account_id": cls.account_selfinvoice.id,
             }
         )
-        self.journal_selfinvoice_extra = journal_model.create(
+        cls.journal_selfinvoice_extra = journal_model.create(
             {
                 "name": "Extra Selfinvoice",
                 "type": "sale",
@@ -74,35 +78,38 @@ class TestInvoiceRC(FatturapaCommon):
             }
         )
 
-    def _create_rc_types(self):
-        rc_type_model = self.env["account.rc.type"]
-        self.rc_type_ita = rc_type_model.create(
+    @classmethod
+    def _create_rc_types(cls):
+        rc_type_model = cls.env["account.rc.type"]
+        cls.rc_type_ita = rc_type_model.create(
             {
                 "name": "RC ITA (selfinvoice)",
                 "method": "selfinvoice",
                 "partner_type": "other",
-                "partner_id": self.env.ref("base.main_partner").id,
-                "journal_id": self.journal_selfinvoice.id,
-                "payment_journal_id": self.journal_reconciliation.id,
-                "transitory_account_id": self.account_selfinvoice.id,
+                "partner_id": cls.env.ref("base.main_partner").id,
+                "journal_id": cls.journal_selfinvoice.id,
+                "payment_journal_id": cls.journal_reconciliation.id,
+                "transitory_account_id": cls.account_selfinvoice.id,
                 "e_invoice_suppliers": True,
             }
         )
 
-    def _create_rc_type_taxes(self):
-        rc_type_tax_model = self.env["account.rc.type.tax"]
-        self.rc_type_tax_ita = rc_type_tax_model.create(
+    @classmethod
+    def _create_rc_type_taxes(cls):
+        rc_type_tax_model = cls.env["account.rc.type.tax"]
+        cls.rc_type_tax_ita = rc_type_tax_model.create(
             {
-                "rc_type_id": self.rc_type_ita.id,
-                "purchase_tax_id": self.tax_22ai.id,
-                "sale_tax_id": self.tax_22vi.id,
+                "rc_type_id": cls.rc_type_ita.id,
+                "purchase_tax_id": cls.tax_22ai.id,
+                "sale_tax_id": cls.tax_22vi.id,
             }
         )
 
-    def _create_fiscal_position(self):
-        model_fiscal_position = self.env["account.fiscal.position"]
-        self.fiscal_position_rc_ita = model_fiscal_position.create(
-            {"name": "RC ITA", "rc_type_id": self.rc_type_ita.id}
+    @classmethod
+    def _create_fiscal_position(cls):
+        model_fiscal_position = cls.env["account.fiscal.position"]
+        cls.fiscal_position_rc_ita = model_fiscal_position.create(
+            {"name": "RC ITA", "rc_type_id": cls.rc_type_ita.id}
         )
 
     def test_00_xml_import(self):

--- a/l10n_it_fatturapa_in_rc/wizard/wizard_import_fatturapa.py
+++ b/l10n_it_fatturapa_in_rc/wizard/wizard_import_fatturapa.py
@@ -50,3 +50,9 @@ class WizardImportFatturapa(models.TransientModel):
             if rc_ita_fp:
                 invoice.fiscal_position_id = rc_ita_fp
         return res
+
+    def _prepareInvoiceLineAliquota(self, credit_account_id, line, nline):
+        retLine = super()._prepareInvoiceLineAliquota(credit_account_id, line, nline)
+        if self._is_in_reverse_charge_line(line):
+            retLine["rc"] = True
+        return retLine

